src/integrations/watchers/indexable-ancestor-watcher.php

```diff
@@ -88,14 +110,119 @@ public function reset_children( $indexable, $indexable_before ) {
 			return false;
 		}
 
-		$children = $this->indexable_repository->get_children( $indexable );
-		foreach ( $children as $child ) {
-			$this->indexable_hierarchy_builder->build( $child );
+		$child_indexable_ids = $this->indexable_hierarchy_repository->find_children( $indexable );
+		$child_indexables    = $this->indexable_repository->find_by_ids( $child_indexable_ids );
+
+		\array_walk( $child_indexables, [ $this, 'refresh_permalink' ] );
 
-			$child->permalink = $this->indexable_helper->get_permalink_for_indexable( $child );
-			$child->save();
+		if ( $indexable->object_type === 'term' ) {
+			$child_indexables_for_term = $this->get_children_for_term( $indexable->object_id, $child_indexables );
+
+			\array_walk( $child_indexables_for_term, [ $this, 'refresh_permalink' ] );
 		}
 
 		return true;
 	}
+
+	/**
+	 * Finds all child indexables for the given term.
+	 *
+	 * @param int         $term_id          Term to fetch the indexable for.
+	 * @param Indexable[] $child_indexables The already known child indexables.
+	 *
+	 * @return array The list of additional child indexables for a given term.
+	 */
+	public function get_children_for_term( $term_id, array $child_indexables ) {
+		// Finds object_ids (posts) for the term.
+		$object_ids = $this->get_object_ids_for_term( $term_id, $child_indexables );
+
+		// Removes the objects that are already present in the children.
+		foreach ( $child_indexables as $child_indexable ) {
+			if ( $child_indexable->object_type !== 'post' ) {
+				continue;
+			}
+
+			$search = \array_search( $child_indexable->object_id, $object_ids, true );
+			if ( $search === false ) {
+				continue;
+			}
+
+			unset( $object_ids[ $search ] );
+		}
+
+		// Finds the indexables for the fetched object_ids.
+		$indexables_by_term = $this->indexable_repository->find_by_multiple_ids_and_type( $object_ids, 'post' , false );
+
+		// Finds the indexables for the posts that attached to the term.
+		$additional_indexable_ids = $this->indexable_hierarchy_repository->find_children_by_ancestor_ids( $object_ids );
```

Comment:
The argument to this query is wrong. This should be the IDs of those indexables, not their object ids.
```suggestion
		$indexable_ids_by_term = \wp_list_pluck( $indexables_by_term, 'id' );
		$additional_indexable_ids = $this->indexable_hierarchy_repository->find_children_by_ancestor_ids( $indexable_ids_by_term );
```